provider/waiter.go

```diff
@@ -0,0 +1,160 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/go-cty/cty"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+)
+
+// Waiter is a simple interface to implement a blocking wait operation
+type Waiter interface {
+	Wait(context.Context) error
+}
+
+// NewResourceWaiter constructs an appropriate Waiter using the supplied waitForBlock configuration
+func NewResourceWaiter(resource dynamic.ResourceInterface, resourceName string, waitForBlock cty.Value) (Waiter, error) {
+	fields := waitForBlock.GetAttr("fields")
+
+	if !fields.IsNull() || fields.IsKnown() {
+		if !fields.Type().IsMapType() {
+			return nil, fmt.Errorf(`"fields" should be a map of strings`)
+		}
+
+		vm := fields.AsValueMap()
+		matchers := []FieldMatcher{}
+		for k, v := range vm {
+			expr := v.AsString()
+			var re *regexp.Regexp
+			if expr == "*" {
+				// NOTE this is just a shorthand so the user doesn't have to
+				// type the expression below all the time
+				re = regexp.MustCompile("(.*)?")
+			} else {
+				var err error
+				re, err = regexp.Compile(expr)
+				if err != nil {
+					return nil, fmt.Errorf("invalid regular expression: %q", expr)
+				}
+			}
+
+			matchers = append(matchers, FieldMatcher{FieldPathToCty(k), re})
+		}
+
+		return &FieldWaiter{
+			resource,
+			resourceName,
+			matchers,
+		}, nil
+	}
+
+	return &NoopWaiter{}, nil
+}
+
+// FieldMatcher contains a cty Path to a field and a regexp to match on it
+type FieldMatcher struct {
+	path         cty.Path
+	valueMatcher *regexp.Regexp
+}
+
+// FieldWaiter will wait for a set of fields to be set,
+// or have a particular value
+type FieldWaiter struct {
+	resource      dynamic.ResourceInterface
+	resourceName  string
+	fieldMatchers []FieldMatcher
+}
+
+// Wait blocks until all of the FieldMatchers configured evaluate to true
+func (w *FieldWaiter) Wait(ctx context.Context) error {
+	return wait(ctx, w.resource, w.resourceName, func(obj cty.Value) (bool, error) {
+		for _, m := range w.fieldMatchers {
+			v, err := m.path.Apply(obj)
+			if err != nil {
+				return false, err
+			}
+
+			if !m.valueMatcher.Match([]byte(v.AsString())) {
```

Comment:
I'm wondering if we should format bools and numbers into string values so we can match on them too. At the moment the provider will blow up if you try and use this on any type other than string.

Reply:
For the moment I am going to format bools and numbers into a string rather than erroring.